nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/ADLSCredentialsControllerService.java

```diff
@@ -74,12 +75,56 @@
         .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
         .build();
 
+    public static final PropertyDescriptor SERVICE_PRINCIPAL_TENANT_ID = new PropertyDescriptor.Builder()
+            .name("service-principal-tenant-id")
+            .displayName("Service Principal Tenant ID")
+            .description("Tenant ID of the Azure Active Directory hosting the Service Principal. The property is required when Service Principal authentication is used.")
+            .sensitive(true)
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
```

Comment:
If this property is considered sensitive, should it support expression language?  Retrieving a sensitive property from flow file attributes would expose the value in provenance events.  This same question applies to the other properties marked as sensitive.

Reply:
That is definitely an issue, but there is an important use case for it -- when you need to shard the data across a lot of storage accounts for performance reasons or you're using storage accounts that may only be known at runtime. When I've seen this, though, the volatile provenance repository was being used so the surface area was small.

It is called out in the property documentation but maybe it could be louder ("certain risks" doesn't sound very scary) or include a more detailed explanation in additional details. What's there now: 

> There are certain risks in allowing the account name to be stored as a flowfile attribute. While it does provide for a more flexible flow by allowing the account name to be fetched dynamically from a flowfile attribute, care must be taken to restrict access to the event provenance data (e.g. by strictly controlling the policies governing provenance for this Processor). In addition, the provenance repositories may be put on encrypted disk partitions.

Reply:
@jfrazee Thanks for providing that background reference to the existing documentation.  Understanding that these new properties fall in the same category of concerns, is it possible to implement the use cases described using Parameter Contexts?  It seems like that would work for retrieving the account information at runtime, but it could make flows a bit more complicated when it is necessary to shard data across storage accounts.  With the goal of moving in a more secure direction, would it be better to avoid introducing new properties supporting expression language here to encourage moving to Parameter Contexts?

Reply:
I don't think the parameter contexts gets at the original desire since you'd have to know in advance which and how many parameters to create, but maybe since it's a new new feature, there's no harm in not allowing EL. We can change our mind later without creating any upgrade issues. The opposite isn't true if we try to remove it later.

The "right" thing to do here could be to say that the use case should use managed identities, or _maybe_ SAS tokens.

Question: how much pain do you think would be induced by rolling this back on Access Keys?

Reply:
The concept of supporting EL for the sensitive `Account Name / Key` came from the Blob credential service (the Blob counterpart of this CS) and from the use case described by @jfrazee. I followed this "pattern" with the new Service Principal properties.

However, there was a bug in this ADLS controller service and ELs were not evaluated at all (this PR would fix it). So at the moment nobody can use the ADLS service with EL and therefore removing EL support for `Account Name / Key` would not arise backward compatibility questions here.

I think we can safely go ahead with removing EL support from all (old and new) sensitive properties if that is preferred. We can add it later if needed.
Implementing a lookup service (like `AzureStorageCredentialsControllerServiceLookup`) is also an option. It can support multiple credentials within the same flow (though it cannot be so dynamic as EL).

Reply:
Thanks for pointing that out @turcsanyip.  I am in favor of removing EL support since it is not currently functional.

Reply:
So it sounds like the only hard decision is whether to remove EL support from the SAS Token property? If created properly they provide fairly fine-grained security so the risk is significantly less than with Access Keys or SPs.

That said, maybe the user should have to work harder to do it. I think a CS could do this though since the client is created on every `onTrigger()`?

Reply:
@jfrazee EL was dysfunctional at all in this CS so `SAS Token` with EL did not work either (sorry, I mentioned only `Account Name / Key` because usually I use those).
ADLS processors were released in 1.12 and I think they are not widely adapted yet. I guess this is the reason why the EL problem has not been noticed so far.

Reply:
Ok. So seems consensus is to not support EL? And we can entertain an alternate CS or lookup service if there's specific demand?

Reply:
Yes, I believe this is the consensus and I'll remove EL support.